pump/server.go

```diff
@@ -505,30 +512,30 @@ func (s *Server) gcBinlogFile() {
 				continue
 			}
 
-			safeTSO, err := s.getSafeGCTSOForDrainers(s.ctx)
-			if err != nil {
-				log.Warn("get save gc tso for drainers failed", zap.Error(err))
-				continue
-			}
-			log.Info("get safe ts for drainers success", zap.Int64("ts", safeTSO))
-
 			millisecond := time.Now().Add(-s.gcDuration).UnixNano() / 1000 / 1000
 			gcTS := int64(oracle.EncodeTSO(millisecond))
-			if safeTSO < gcTS {
-				gcTS = safeTSO
-			}
-			log.Info("send gc request to storage", zap.Int64("ts", gcTS))
-			s.storage.GCTS(gcTS)
+
+			alertGCMS := millisecond + earlyAlertGC.Nanoseconds()/1000/1000
+			alertGCTS := int64(oracle.EncodeTSO(alertGCMS))
+
+			realGCTS := s.storage.GCTS(gcTS)
+			log.Info("send gc request to storage", zap.Int64("request gc ts", gcTS), zap.Int64("real gc ts", realGCTS))
+
+			atomic.StoreInt64(&s.gcTS, realGCTS)
+			// detect where drainer's checkpoint is old than pump gc pointer
+			// it also means binlog events is loss for this drainer
+			s.getSafeGCTSOForDrainers(s.ctx, alertGCTS)
 		}
 	}
 }
 
-func (s *Server) getSafeGCTSOForDrainers(ctx context.Context) (int64, error) {
+func (s *Server) getSafeGCTSOForDrainers(ctx context.Context, gcTS int64) (int64, error) {
```

Comment:
I'm afraid it's not a good idea to reuse `getSafeGCTSOForDrainers` in this way. It does something that the original name doesn't imply and it changes the interface.

I think it would be more clear if we just extract another function like `getOnlineDrainers` and then reuse that function.

Reply:
what does `getOnlineDrainers` do? I just want a function do
1. query drainer status
2. for all drainer, and check it's max commit ts

so I may need a similar function as getSafeGCTSOForDrainers

Reply:
If we have a `getOnlineDrainers` function, it can be reused in both `getSafeGCTSOForDrainers` and the function you want here, no duplication code needed and the original simple interface of `getSafeGCTSOForDrainers` doesn't have to be changed.

Reply:
This is really not necessary, because you always define it when you need it, and there are not a few lines of code at all.